### PR TITLE
feat: scifi: Direct controller for Alloy DB Cluster

### PR DESCRIPTION
**Pull request by anhdle-sso**
### Change description

Fixes #

### Tests you have done

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [x] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: pkg/controller/direct/alloydb/cluster_controller.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+		log.V(0).Info("change detected: displayName")
+		updateMask.Paths = append(updateMask.Paths, "displayName")
+	}
+	if a.desired.ClusterType != a.actual.ClusterType {
+		log.V(0).Info("change detected: clusterType")
+		updateMask.Paths = append(updateMask.Paths, "clusterType")
+	}
+	if !reflect.DeepEqual(a.desired.EncryptionConfig, a.actual.EncryptionConfig) {
+		log.V(0).Info("change detected: encryption_config")
+		updateMask.Paths = append(updateMask.Paths, "encryption_config")
```
Comment (justinsb) [inline pkg/controller/direct/alloydb/cluster_controller.go:241]: BTW, I'm trying to figure out if we've documented anywhere what our behaviour should be here.  mockgcp uses grpc-gateway, which is very strict about the names, but I think real-gcp allows either proto or json format.  If you find any docs, please LMK!

File: pkg/controller/direct/alloydb/mapping.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+	if in == nil {
+		return nil
+	}
+	out := &api.UserPassword{
+		User: ValueOf(in.User),
+	}
+	if in.Password.Value != nil {
+		out.Password = ValueOf(in.Password.Value)
+	} else {
+		out.Password = in.Password.ValueFrom.SecretKeyRef.Key
```
Comment (justinsb) [inline pkg/controller/direct/alloydb/mapping.go:159]: Is this right?  You can add errors to `ctx *MapContext` if that's more appropriate.

One thing I'm not certain about is when we should resolve these i.e. whether we resolve them on demand during the mapping, or whether we try to resolve them all up-front and (here) replace ValueFrom with  Value.  For ProjectRefs, we would replace Name with External. etc

File: pkg/controller/direct/alloydb/mapping.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+		DatabaseVersion:      LazyPtr(in.DatabaseVersion),
+		EncryptionInfo:       EncryptionInfo_KRMFromApi(ctx, in.EncryptionInfo),
+		MigrationSource:      MigrationSource_KRMFromApi(ctx, in.MigrationSource),
+		Name:                 LazyPtr(in.Name),
+		Uid:                  LazyPtr(in.Uid),
+	}
+	return out
+}
+
+func BackupSource_KRMFromApi(ctx *MapContext, in *api.BackupSource) []krm.ClusterBackupSourceStatus {
```
Comment (justinsb) [inline pkg/controller/direct/alloydb/mapping.go:180]: Ouch ... do we have the cardinality wrong in our API?  Still, at least it looks like we have it wrong the less-problematic way.

**Comment by justinsb** [comment]
A few comments, but nothing that can't be done in follow-on (as this is feature-flagged)  I'll add hold, but feel free to cancel the hold when you've seen the comments.

/approve
/lgtm
/hold

**Comment by acpana** [comment]
hm, it seems this included some out of date interfaces for the direct controller

**Pull request merged**